src/adaptation/precond.jl

```diff
@@ -29,38 +30,42 @@ function reset!(nc::NaiveVar{T}) where {T<:AbstractFloat}
     nc.S = Vector{Vector{T}}()
 end
 
-function get_var(nc::NaiveVar{T})::Vector{T} where {T<:AbstractFloat}
+function get_var(nc::NaiveVar{T}) where {T<:AbstractFloat}
     @assert nc.n >= 2 "Cannot get variance with only one sample"
     return Statistics.var(nc.S)
 end
 
 # Ref： https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_var_estimator.hpp
-mutable struct WelfordVar{T<:AbstractFloat,AT<:AbstractVector{T}} <: VarEstimator{T}
+mutable struct WelfordVar{T<:AbstractFloat, AT<:AbstractVector{T}} <: VarEstimator{T}
     n :: Int
     μ :: AT
     M :: AT
 end
 
-WelfordVar(d::Int) = WelfordVar(0, zeros(d), zeros(d))
+WelfordVar(::Type{T}, d::Int) where {T} = WelfordVar(0, zeros(T, d), zeros(T, d))
+WelfordVar(d::Int) = WelfordVar(Float64, d)
 
-function reset!(wv::WelfordVar{T,AT}) where {T<:AbstractFloat,AT<:AbstractVector{T}}
+function reset!(wv::WelfordVar{T, AT}) where {T<:AbstractFloat, AT<:AbstractVector{T}}
     wv.n = 0
     wv.μ .= zero(T)
     wv.M .= zero(T)
 end
 
 function add_sample!(wv::WelfordVar, s::AbstractVector)
     wv.n += 1
-    δ = s .- wv.μ
-    wv.μ .+= δ ./ wv.n
-    wv.M .+= δ .* (s .- wv.μ)
+    @unpack μ, M, n = wv
+    for i in eachindex(s)
+        δ = s[i] - μ[i]
+        μ[i] += δ / n
+        M[i] += δ * (s[i] - μ[i])
```

Comment:
Why switching from broadcasting to loop? Is this for readability or for efficiency?

Reply:
Efficiency to avoid allocating `δ`.